packages/hooking/hook.ts

```diff
@@ -14,33 +14,85 @@
  * limitations under the License.
  */
 
-/*eslint @typescript-eslint/no-explicit-any: 0 */
+// Stores package names and function names [packageName -> [fnName, ...], ...].
+// This structure is used to keep track of all functions seen during instrumentation and execution of the fuzzing run,
+// to determine which hooks have been applied, are available, and have not been applied.
+export class TrackedHooks {
+	hooks: Map<string, Set<string>>;
+	constructor() {
+		this.hooks = new Map();
+	}
+
+	add(pkg: string, target: string) {
+		if (!this.hooks.has(pkg)) {
+			this.hooks.set(pkg, new Set());
+		}
+		this.hooks.get(pkg)?.add(target);
+	}
+
+	has(pkg: string, target: string) {
+		if (!this.hooks.has(pkg)) {
+			return false;
+		}
+		return this.hooks.get(pkg)?.has(target);
+	}
+
+	serialize(): TrackedHookData[] {
+		const result: TrackedHookData[] = [];
+		for (const [pkg, targets] of [...this.hooks].sort()) {
+			for (const target of [...targets].sort()) {
+				result.push({ pkg: pkg, target: target });
+			}
+		}
+		return result;
+	}
+
+	clear() {
+		this.hooks.clear();
+	}
+
+	get size() {
+		let size = 0;
+		for (const targets of this.hooks.values()) {
+			size += targets.size;
+		}
+		return size;
+	}
+}
+
+export interface TrackedHookData {
+	target: string;
+	pkg: string;
+}
 
 class hookTracker {
```

Comment:
Classes should start with an uppercase character.

Reply:
I am not the originator, hehe :sweat_smile:  Changed.